packages/devextreme/testing/tests/DevExpress.ui.widgets/menu.tests.js

```diff
@@ -380,6 +386,199 @@ QUnit.module('Menu rendering', {
+    });
+
+    QUnit.test('Nested submenu should be positioned to a clicked item', function(assert) {
+        if(!isDeviceDesktop(assert)) {
+            return;
+        }
+
+        const menu = createMenuInWindow({
+            items: [{
+                text: 'item 1',
+                items: [{
+                    text: 'item 11',
+                    items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
+                }],
+            }],
+            showFirstSubmenuMode: 'onClick',
+            showSubmenuMode: { name: 'onHover', delay: 0 },
+        });
+        const $rootItem = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        $($rootItem).trigger('dxclick');
+        this.clock.tick(0);
+
+        const submenu = getSubMenuInstance($rootItem);
+        const $menuItem = $($(submenu._overlay.content()).find(`.${DX_MENU_ITEM_CLASS}`).first());
+
+        $(submenu.itemsContainer()).trigger($.Event('dxhoverstart', { target: $menuItem.get(0) }));
+        $($menuItem).trigger('dxpointermove');
+        this.clock.tick(0);
+
+        const $nestedSubmenu = $($(submenu._overlay.content()).find(`.${DX_SUBMENU_CLASS}`).eq(1));
+        const $nestedItemsContainer = $nestedSubmenu.find(`.${DX_CONTEXT_MENU_ITEMS_CONTAINER_CLASS}`).eq(0);
+
+        assert.roughEqual($nestedItemsContainer.offset().top, $menuItem.offset().top, .1, 'Nested submenu positioned to a clicked item');
+        assert.roughEqual(
+            $nestedSubmenu.outerHeight(), $(window).height() - $nestedSubmenu.offset().top - BORDER_WIDTH,
+            .1, 'Nested submenu uses all available space');
+    });
+
+    QUnit.test('Flipping root submenu', function(assert) {
+        const menu = createMenuInWindow({
+            items: [{
+                text: 'item 1',
+                items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
+            }],
+            showFirstSubmenuMode: 'onClick',
+        });
+        const $item1 = $(menu.element).find(`.${DX_MENU_ITEM_CLASS}`).eq(0);
+
+        menu.element.css('top', 10000 + $(window).height() - 50);
+        $($item1).trigger('dxclick');
+        this.clock.tick(0);
+
+        const $submenu = $(`.${DX_SUBMENU_CLASS}`);
+
+        assert.ok($submenu.find(`.${DX_SCROLLVIEW_CONTENT_CLASS}`).height() > $(window).height(), 'total height of submenu exceeds the window height');
+        assert.roughEqual($submenu.offset().top, 0, .1, 'submenu flipped to top');
+        assert.roughEqual($submenu.outerHeight(), $(window).height() - 50, .1, 'menu uses all available space');
+    });
+
+    QUnit.test('Flipping 2nd level submenu', function(assert) {
+        if(!isDeviceDesktop(assert)) {
+            return;
+        }
+
+        const menu = createMenuInWindow({
+            items: [{
+                text: 'item 1',
+                items: [{
+                    text: 'item 11',
+                    items: (new Array(99)).fill(null).map((_, idx) => ({ text: idx })),
```

Comment:
Don't u wanna to union all these tests into single module to unify some things like items creation, submenu getting function, etc?